### PR TITLE
Improve rules, rule testing, and bump dagon

### PR DESCRIPTION
**Pull request by johnynek**
This seems to fix an issue with an infinite loop I was seeing that we were interpreting as it taking a long time.

I think when randomly making rules: one would change the graph in one way, the next would change it back, creating an infinite loop.

I seem to fix it by removing a rule from random selection. I also check that the standard rules don't loop infinitely.

### CODE REVIEW

File: scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala [hunk 1]
```diff
@@ -545,12 +527,28 @@ object OptimizationRules {
...
   }
 
+  /**
+   * This rule is important in that it allows us to reduce
+   * the number of nodes in the graph, which is helpful to speed up rule application
+   */
   object ComposeDescriptions extends PartialRule[TypedPipe] {
+    def combine(descs1: List[(String, Boolean)], descs2: List[(String, Boolean)]): List[(String, Boolean)] = {
+      val combined = descs1 ::: descs2
+      val unordered = combined.collect { case (d, true) => d }.toSet
```
Comment (ianoc-stripe) [inline scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala]: this is unused?

**Comment by ianoc-stripe** [comment]
lgtm

File: scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala (hunk 1 above)
Comment (johnynek) [reply scalding-core/src/main/scala/com/twitter/scalding/typed/OptimizationRules.scala]: Nope. I somehow forgot it and implemented without it below. Good catch.

**Comment by non** [comment]
lgtm 👍

**Comment by johnynek** [comment]
the only test that failed is the one that is fixed by #1805.

Merging since that one should fix it up.

**Pull request merged**